.github/workflows/_setup.yml

```diff
@@ -0,0 +1,73 @@
+# For more information see:
+# - https://docs.github.com/en/actions/using-workflows/reusing-workflows
+# - https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations
+
+# Note:
+# The workflow can be referenced in other workflows like:
+# - ngxs/store/.github/workflows/_setup.yml@master
+# - ./.github/workflows/_setup.yml
+
+name: setup
+
+on:
+  workflow_call:
+    inputs:
+      NODE_VERSION:
+        required: true
+        type: string
+      GITHUB_SHA:
+        required: true
+        type: string
+      GITHUB_REF:
+        required: true
+        type: string
+
+jobs:
+  setup:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          ref: ${{ inputs.GITHUB_SHA }}
+
+      - name: Don't save Bash session history
+        run: unset HISTFILE
+
+      - name: Use Node.js ${{ inputs.NODE_VERSION }}
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ inputs.NODE_VERSION }}
+          check-latest: true
+
+      - name: Configure kernel (increase watchers)
+        run: echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p
+
+      - name: Get variables (branch name, commit hash, yarn cache directory)
+        id: get-variables
+        run: |
+          echo "::set-output name=shortref::${inputs.GITHUB_REF#refs/*/}";
+          echo "::set-output name=commitsha::$(echo ${inputs.GITHUB_SHA})";
+          echo "::set-output name=yarncachedir::$(yarn cache dir)";
+
+      - name: Workspace cache
+        uses: actions/cache@v3
+        id: workspace-cache
+        env:
+          cache-name: workspace-cache
+        with:
+          path: |
+            ~/.cache
+            ./node_modules
+            ./@ngxs
```

Comment:
@markwhitfeld I think in this particular case the chance of something going wrong with the artifacts is extremely low.
Here's why:
1. when the workflow (trunk or release) is kicked off, the first job is `*-build` (`trunk-build` or `release-build`) <- `*-build` jobs refresh `workspace-cache` by rebuilding all packages and saving it to the cache.
2. Then several more jobs are executed: `*-test`, `*-integration-test`, `*-bundlesize`.
3. Then the `*-publish` job is executed. The `publish` job uses build artifacts that were generated by the `*-build` job (see point 1 of this list).

A possible attack surface is in between of `build` and `publish`, it's point 2 in the list above.
The attacker should be very persistent and determined to mess up the publish step.
If we assume that the workspace artifacts (including the build artifacts) are corrupted before `*-build`, they will be fixed after the `*-build` job completion.

I think, with the same probability someone could tap into the runner and corrupt the build artifacts. Of course there is a probability of this, but it tends to zero.

If you want an additional layer of protection to be sure that the build artifacts were not changed after caching and before publishing, we can hash the content of each directory with the build artifacts and save hashing results as part of the `workspace-cache`. The `_check-build-artifacts.yml` workflow can be configured to verify hashes in addition to ensuring that the directories exist.

Reply:
Thank you for your very well-thought-out response.
I think that it is good to bear this in mind, but I don't think it is essential right now.
I think that the simplest approach is to use the cache so that it is available to the next job, and then in the publish step, we just build it again. It is repeating the work, but everything preceding the publish leveraged the built files as part of the validation leading up to the publish. It is OK if the publish repeats a number of steps to be safer because it is on the tail end of the workflow and will not slow the PR workflow down.